builtin/clone.mbt

```diff
@@ -0,0 +1,261 @@
+  inspect!(xs, content="[99, 2, 3, 4, 5]")
+  inspect!(ys, content="[1, 2, 3, 4, 5]")
+}
+
+test "FixedArray[Ref[Int]]::clone" {
+  let xs = ([1, 2, 3, 4, 5] : FixedArray[Int]).map_fixedarray(fn(x) {
+    { val: x }
+  })
+  let ys = xs.clone()
+  xs[0].val = 99
+  inspect!(xs, content="[{val: 99}, {val: 2}, {val: 3}, {val: 4}, {val: 5}]")
+  inspect!(ys, content="[{val: 1}, {val: 2}, {val: 3}, {val: 4}, {val: 5}]")
+}
+
+///|
+impl[T : Clone] Clone for Ref[T] with mutability() { Mutable }
+
+///|
+impl[T : Clone] Clone for Ref[T] with clone(self) {
+  match T::mutability() {
+    Mutable => { val: T::clone(self.val) }
+    Immutable => { val: self.val }
+  }
+}
+
+test "Ref[Int]::clone" {
+  let x = { val: 0 }
+  let y = x.clone()
+  x.val = 10
+  inspect!(x, content="{val: 10}")
+  inspect!(y, content="{val: 0}")
+}
+
+///|
+impl Clone for Bytes with mutability() { Mutable }
+
+///|
+impl Clone for Bytes with clone(self) { self.copy() }
+
+test "Bytes::clone" {
+  let xs = b"\x12\x34"
+  let ys = xs.clone()
+  xs[0] = b'\x00'
+  let xss =
+    #|b"\x00\x34"
+  let yss =
+    #|b"\x12\x34"
+  inspect!(xs, content=xss)
+  inspect!(ys, content=yss)
+}
+
+// immutable containers
+
+///|
+impl[T : Clone] Clone for T? with mutability() { T::mutability() }
+
+///|
+impl[T : Clone] Clone for T? with clone(self) {
+  match T::mutability() {
+    Mutable =>
+      match self {
+        Some(x) => Some(T::clone(x))
+        None => None
+      }
+    Immutable => self
+  }
+}
+
+test "Ref[Int]?::clone" {
+  let x = Some({ val: 0 })
+  let y = x.clone()
+  x.unwrap().val = 99
+  inspect!(x, content="Some({val: 99})")
+  inspect!(y, content="Some({val: 0})")
+}
+
+///|
+impl[T : Clone, E : Clone] Clone for Result[T, E] with mutability() {
+  match (T::mutability(), E::mutability()) {
+    (Immutable, Immutable) => Immutable
+    _ => Mutable
+  }
+}
+
+///|
+impl[T : Clone, E : Clone] Clone for Result[T, E] with clone(self) {
+  match T::mutability() {
+    Mutable =>
+      match self {
+        Ok(x) => Ok(T::clone(x))
+        Err(x) => Err(E::clone(x))
+      }
+    Immutable => self
+  }
+}
+
+test "Result[Ref[Int],Ref[String]]::clone" {
+  let x : Result[Ref[Int], Ref[String]] = Ok({ val: 1 })
+  let y = x.clone()
+  match x {
+    Ok(x) => x.val = 99
+    Err(x) => x.val = "OvO"
+  }
+  inspect!(x, content="Ok({val: 99})")
+  inspect!(y, content="Ok({val: 1})")
+  let x : Result[Ref[Int], Ref[String]] = Err({ val: "mooncake" })
+  let y = x.clone()
+  match x {
+    Ok(x) => x.val = 99
+    Err(x) => x.val = "OvO"
+  }
+  inspect!(x, content="Err({val: \"OvO\"})")
+  inspect!(y, content="Err({val: \"mooncake\"})")
+}
+
+///|
+fn get_mutability[T : Clone](_ : T) -> Mutability {
+  T::mutability()
+}
+
+test "mutability" {
+
+  // immutable 
+  let oi = Some(0)
+  inspect!(get_mutability(oi), content="Immutable")
+  let ris : Result[Int, String] = Ok(32)
+  inspect!(get_mutability(ris), content="Immutable")
+  // TODO, we need to test reference equal
```

Comment:
we have a function `physical_equal` but be aware that its behavior is backend-dependent.